library/core/src/mem/mod.rs

```diff
@@ -642,24 +652,19 @@ pub unsafe fn zeroed<T>() -> T {
 /// is special in that the compiler knows that it does not have a fixed value.
 /// This makes it undefined behavior to have uninitialized data in a variable even
 /// if that variable has an integer type.
-/// (Notice that the rules around uninitialized integers are not finalized yet, but
-/// until they are, it is advisable to avoid them.)
 ///
 /// [uninit]: MaybeUninit::uninit
 /// [assume_init]: MaybeUninit::assume_init
 /// [inv]: MaybeUninit#initialization-invariant
 #[inline(always)]
 #[rustc_deprecated(since = "1.39.0", reason = "use `mem::MaybeUninit` instead")]
 #[stable(feature = "rust1", since = "1.0.0")]
-#[allow(deprecated_in_future)]
-#[allow(deprecated)]
 #[rustc_diagnostic_item = "mem_uninitialized"]
 pub unsafe fn uninitialized<T>() -> T {
     // SAFETY: the caller must guarantee that an unitialized value is valid for `T`.
-    unsafe {
-        intrinsics::assert_uninit_valid::<T>();
-        MaybeUninit::uninit().assume_init()
-    }
+    // Because an uninitialized value does not guarantee any specific bit
+    // representation, it is therefore no less safe to return a zeroed value.
+    unsafe { zeroed::<T>() }
```

Comment:
Leaving this as a separate comment so that it does not get forgotten: when compiled with `cfg(miri)`, the code should be the same as before (at least the `MaybeUninit::uninit().assume_init()` part, the panic is less relevant here) so that Miri can still detect incorrect uses of `mem::uninitialized`.

Reply:
I would also suggest leaving the memory uninitialized when MemorySanitizer is enabled, i.e., when `#[cfg(sanitize = "memory")]`. The test suite uses `MaybeUninit`, so this was unfortunately not caught automatically.

Reply:
I've implemented both of these in the most recent push. I tried for a bit to get my locally-built copy of miri to work with my locally-built copy of rustc, but gave up after hitting ICEs, so I haven't tested it.

Reply:
FWIW, there's instructions for this at https://github.com/rust-lang/miri/blob/master/CONTRIBUTING.md#building-miri-with-a-locally-built-rustc -- but this should be easy enough to check by looking at the code.

Thanks!